pkg/generate/code/compare_test.go

```diff
@@ -0,0 +1,110 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	 http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package code_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/aws-controllers-k8s/code-generator/pkg/generate/code"
+	"github.com/aws-controllers-k8s/code-generator/pkg/testutil"
+)
+
+func TestCompareResource_S3_Bucket(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	g := testutil.NewGeneratorForService(t, "s3")
+
+	crd := testutil.GetCRDByName(t, g, "Bucket")
+	require.NotNil(crd)
+
+	expected := `
+	if ackcompare.HasNilDifference(a.ko.Spec.ACL, b.ko.Spec.ACL) {
+		delta.Add("Spec.ACL", a.ko.Spec.ACL, b.ko.Spec.ACL)
+	} else {
+		if *a.ko.Spec.ACL != *b.ko.Spec.ACL {
```

Comment:
While testing the s3 controller (generated with the new compare methods) i observed few runtime panics, and the reason is that we don't handle cases where both of `a.ko.Spec.*` and `a.ko.Spec.*` are nil pointers. Calling "valueof" of a nil pointer causes  panic.

Reply:
quick fix (Probably not the right way to fix it): https://github.com/A-Hilaly/ack-code-generator/commit/f7873768b73ae05b2147bd2fd025bbdb8ca9569c :)

Reply:
Another excellent catch, @A-Hilaly! I went with a route that generated:

```go
} else if a.ko.Spec.Name != nil && b.ko.Spec.Name != nil {
```

instead of an additional indent-level if statement.

Reply:
Great! 👍